### PR TITLE
Describe and clarify some of the control flow in core.py

### PR DESCRIPTION
**Pull request by Zalathar**
In the past I've had some trouble navigating the control flow in `given` and `StateForActualGivenExecution`, so this time I wrote up some of my notes as comments.

Hopefully this will save others (and my future self) some confusion.

---

Most of the changes are comments and docstrings, but I also made a few small code changes to make things clearer in key places. There shouldn't be any functional changes.

---

All of these comments are based on my own reading of the code, so while I've tried to be accurate, mistakes are certainly possible.

### CODE REVIEW

**Comment by Zac-HD** [comment]
This looks really helpful - thanks @Zalathar!

File: hypothesis-python/src/hypothesis/core.py [hunk 1]
```diff
@@ -536,9 +555,12 @@ def test(*args, **kwargs):
                 return result
 
         def run(data):
+            # Set up dynamic context needed by a single test run.
             with local_settings(self.settings):
```
Comment (DRMacIver) [inline hypothesis-python/src/hypothesis/core.py]: I wonder if we still need this. We no longer actually support nesting given, right?

(This is thinking out loud I'm not expecting you to do anything with it in this PR)

File: hypothesis-python/src/hypothesis/core.py [hunk 2]
```diff
@@ -936,10 +1011,17 @@ def wrapped_test(*arguments, **kwargs):
...
                     )
 
+            # There was no @reproduce_failure, so start by running any explicit
+            # examples from @example decorators.
+
             execute_explicit_examples(
                 test_runner, test, wrapped_test, settings, arguments, kwargs
             )
 
+            # All of the explicit examples passed (if any), so the next step
```
Comment (DRMacIver) [inline hypothesis-python/src/hypothesis/core.py]: nit: The placement of the (if any) in this sentence is weird. Maybe "All of the explicit examples (if there were any) passed"?

File: hypothesis-python/src/hypothesis/core.py [hunk 3]
```diff
@@ -813,13 +876,20 @@ def run_test_with_generator(test):
...
-            generator_kwargs[name] = strategy
+        # Because the argument check succeeded, we can convert @given's
+        # positional arguments into keyword arguments for simplicity.
+        if generator_arguments:
+            assert not generator_kwargs
+            for name, strategy in zip(
+                reversed(original_argspec.args), reversed(generator_arguments)
+            ):
+                generator_kwargs[name] = strategy
+            generator_arguments = ()
```
[the remark this answers is not in the document]
Comment (aarchiba) [reply hypothesis-python/src/hypothesis/core.py]: Defensive programming? This way the cases of positional and keyword arguments are indistinguishable afterwards? One could be more emphatic by unconditionally deleting the variable generator_arguments.

[the remark this answers is not in the document]
Comment (Zalathar) [reply hypothesis-python/src/hypothesis/core.py]: Yes, it’s entirely defensive/communicative, and has no actual effect.

The background is that I was confused by the fact that we pass two copies of the positional arguments to the argument processor (args and converted kwargs), and then ignore one of them. So I wanted to clarify that the args have been *moved* into kwargs.

In hindsight, I should probably just `del` the positional args instead.

File: hypothesis-python/src/hypothesis/core.py [hunk 4]
```diff
@@ -510,7 +518,18 @@ def __init__(self, test_runner, search_strategy, test, settings, random, had_see
...
+        self, data, print_example=False, is_final=False, expected_failure=None
+    ):
+        """Run the test function once, using ``data`` as input.
+
+        If the test raises an exception, it will propagate through to the
+        caller of this method. Depending on its type, this could represent
+        an ordinary test failure, or a fatal error, or a control exception.
+
+        If this method returns normally, then the outcome of the test is
+        indicated by ``data.status``.
```
[the remark this answers is not in the document]
Comment (Zalathar) [reply hypothesis-python/src/hypothesis/core.py]: The idea was that a normal return usually means success, but occasionally means a suppressed control exception (e.g. from overrun) that has nevertheless left data in an unsuccessful state.

I’ll change this to say so more directly.

**Comment by Zalathar** [comment]
Looks like some kind of unrelated flaky failure: https://dev.azure.com/HypothesisWorks/Hypothesis/_build/results?buildId=2917

File: hypothesis-python/src/hypothesis/core.py (hunk 1 above)
Comment (aarchiba) [reply hypothesis-python/src/hypothesis/core.py]: given definitely refuses to wrap itself.

Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/core.py]:
```python
# Banned!
@given(x=...)
@given(y=...)
def test(x, y): ...

# Allowed!
@given(x=...)
def test(x):
    @given(y=...)
    def inner(y): ...
```

I think we want to continue allowing the latter form, so I'd leave this as-is.

Comment (DRMacIver) [reply hypothesis-python/src/hypothesis/core.py]: I definitely don't think we *want* to continue allowing the latter form, but I guess you're right that we've not currently deprecated it.

**Pull request merged**